clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -326,58 +548,60 @@ class device_vector {
   }
   void insert(iterator position, size_type n, const T &x) {
     if (position == end()) {
-      resize(size() + n);
-      ::std::fill(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                end() - n, end(), x);
+      reserve(size() + n);
+      device_allocator_traits<Allocator>::uninitialized_fill_n(
+          _alloc, _storage + size(), n, x);
+      _size += n;
     } else {
       auto i_n = ::std::distance(begin(), position);
       // allocate temporary storage
       auto m = ::std::distance(position, end());
       // will throw if position is not inside active vector
-      auto tmp = _alloc.allocate(m);
+      auto tmp = alloc_traits::allocate(_alloc, m);
       // copy remainder
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                position, end(), tmp);
-
+      device_allocator_traits<Allocator>::uninitialized_device_copy_n(
+          _alloc, position, m, tmp);
       resize(size() + n);
       // resizing might invalidate position
       position = begin() + position.get_idx();
-
-      ::std::fill(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                position, position + n, x);
-
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                tmp, tmp + m, position + n);
-      _alloc.deallocate(tmp, m);
+      ::std::fill(
+          oneapi::dpl::execution::make_device_policy(get_default_queue()),
+          position, position + n, x);
+      ::std::copy(
+          oneapi::dpl::execution::make_device_policy(get_default_queue()), tmp,
+          tmp + m, position + n);
+      device_allocator_traits<Allocator>::destroy_n(_alloc, tmp, m);
+      alloc_traits::deallocate(_alloc, tmp, m);
     }
   }
   template <typename InputIterator>
   void
   insert(iterator position, InputIterator first,
          typename ::std::enable_if<internal::is_iterator<InputIterator>::value,
-                                 InputIterator>::type last) {
+                                   InputIterator>::type last) {
     auto n = ::std::distance(first, last);
     if (position == end()) {
       resize(size() + n);
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                first, last, end());
+      // unsafe to call on device as we dont know the InputIterator type
```

Comment:
I think we can use `uninitialized_copy_n` here.

Reply:
done.  I looked at changing some of the other `resize` to `reserve` and `uninitialized_x`, but the others are not clear wins like this one seems to be.